src/redis-cli.c

```diff
@@ -9820,6 +10013,571 @@ void testHintSuite(char *filename) {
     exit(fail);
 }
 
+/*------------------------------------------------------------------------------
+ * Keystats
+ *--------------------------------------------------------------------------- */
+
+/* key length distribution. */
+
+typedef struct size_dist_entry {
+    unsigned long long size;
+    unsigned long long count;
+} size_dist_entry;
+
+typedef struct size_dist {
+    unsigned long long total_count;
+    unsigned long long total_size;
+    unsigned long long max_size;
+    size_dist_entry *size_dist;
+} size_dist;
+
+/* distribution is an array initialized with last element {0, 0}                  */
+/* for instance: size_dist_entry distribution[] = { {32, 0}, {256, 0}, {0, 0} }; */
+static void sizeDistInit(size_dist *dist, size_dist_entry *distribution) {
+    dist->max_size = 0;
+    dist->total_count = 0;
+    dist->total_size = 0;
+    dist->size_dist = distribution;
+}
+
+static void addSizeDist(size_dist *dist, unsigned long long size) {
+    dist->total_count++;
+    dist->total_size += size;
+
+    if (size > dist->max_size) {
+        dist->max_size = size;
+    }
+
+    int j;
+    for (j=0; size > dist->size_dist[j].size && dist->size_dist[j].size; j++);
```

Comment:
check reaching the end in advance?
```suggestion
    for (int j=0; dist->size_dist[j].size && size > dist->size_dist[j].size;  j++);
```

Reply:
True, we could check we reached the end first.
Though, we need `int j;` outside the `for` as is it used to add to the count on the next line:
`dist->size_dist[j].count++;`.

Reply:
is it much better? this make distribution more readable and there is no need to check `dist->size_dist[j].size`.
```diff
diff --git a/src/redis-cli.c b/src/redis-cli.c
index 032250fc7..58ed6988d 100644
--- a/src/redis-cli.c
+++ b/src/redis-cli.c
@@ -10444,7 +10444,7 @@ static void keyStats(long long memkeys_samples, unsigned long long cursor, unsig
         {16*1024*1024, 0},         /*  16 MB                                                  */
         {128*1024*1024, 0},        /* 128 MB                                                  */
         {512*1024*1024, 0},        /* 512 MB (max String size)                                */
-        {0, 0},                    /* Sizes above the last entry                              */
+        {UINT64_MAX, 0},           /* Sizes above the last entry                              */
     };
     sizeDistInit(&key_length_dist, distribution);
```

Reply:
That might work.
We need to change the comment above `sizeDistInit()` to replace {0, 0}.
In addition, we need to change `displayKeyStatsLengthDist()` to use UINT64_MAX as `for (int i=0; dist->size_dist[i].size; i++)` does not work in that case.
Let me test it.

Reply:
The latest commit is using UINT64_MAX.